### PR TITLE
Logging stops

### LINKED ISSUE #212: Log jamming - Sensor logs seem to get bottlenecked
**Describe the bug**
Unsure exactly what has happened here but my tablet is no longer receiving logs from the sensors. 

image

Tried restarting app, removing a sensor, restarting tablet and waiting more than 24 hours but no luck. 

The only way I could get them to update was to open the Blue Maestro app, which seems to trigger an update. 

Interestingly 2 of the sensors on the Blue Maestro app are showing 6000 logs...

**To Reproduce**
1. Unsure - connect 6 sensors, set them to 1 min logging and close app for multiple (6?) days?
2. Turn app back on and see that logs are stuck?

**Expected behavior**

**Smartphone (please complete the following information):**
 - Device: Lenovo Tab X505N
 - OS: Android 10
 - Version 0.5.0-rc0

**Additional context**
Thought it might be the Laird sensor, so removed that, but still seems to be happening?

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #212 

The blue maestro sensors are not reporting connection status correctly:
- deviceIsConnected? `{ deviceIsConnected: true }`
- try disconnecting `[BleError: Device D7:D6:67:E0:02:34 is not connected]`
- trying to connect: `[BleError: Device ? is already connected]`

to work around this - am now disconnecting blue maestro while catching and ignoring errors, then connecting. 

Another issue is that the uuid generation is throwing an error `uuid.v1(): Can\'t create more than 10M uuids/sec` when being called within a `map` loop if there are a lot of logs. Am pausing for a millisecond if this happens to allow the uuid generation to continue.

### CODE REVIEW

File: src/common/services/UtilService/UtilService.ts [hunk 1]
```diff
@@ -5,10 +5,28 @@ import packageJson from '~/../../package.json';
...
+      // catch the error "uuid.v1(): Can\'t create more than 10M uuids/sec"
+      if (e.message.includes('more than 10M uuids/sec')) {
+        // pause to allow the uuid to be generated and try again, throw error if it fails again
+        const start = new Date().getTime();
+        for (let i = 0; i < 1e7; i++) {
+          if (new Date().getTime() - start > PAUSE_IN_MILLISECONDS) {
+            break;
+          }
+        }
+        return generateUUID.v1() as string;
```
Comment (josh-griffin) [inline src/common/services/UtilService/UtilService.ts:25]: oh gosh I should have looked at the implementation before using https://github.com/eugenehp/react-native-uuid/blob/36e31d95bd46d5d86f1a90a7e5d1d192fdb7e122/src/v1.ts#L93 :(

File: .yalc/msupply-ble-service/src/Bluetooth/BleService.ts [hunk 2]
```diff
@@ -79,9 +79,20 @@ export class BleService {
...
+    // to work around this, we disconnect the device, ignoring any errors, before connecting again
+    if (device.deviceType === BLUE_MAESTRO) {
+      try {
+        await this.manager.cancelDeviceConnection(device.id);
+      } catch {
+        // ignore error
+      }
+    } else {
+      const deviceIsConnected = await this.manager.isDeviceConnected(device.id);
+      this.logger.info('deviceIsConnected?', { deviceIsConnected });
```
Comment (josh-griffin) [inline .yalc/msupply-ble-service/src/Bluetooth/BleService.ts:95]: keep this still?

File: .yalc/msupply-ble-service/src/Bluetooth/BleService.ts [hunk 3]
```diff
@@ -79,9 +79,20 @@ export class BleService {
...
     this.logger.info('connectAndDiscoverServices', { deviceDescriptor });
     const device = this.utils.deviceDescriptorToDevice(deviceDescriptor);
-    const deviceIsConnected = await this.manager.isDeviceConnected(device.id);
-    this.logger.info('deviceIsConnected?', { deviceIsConnected });
-    if (deviceIsConnected) {
+    // the Blue Maestro devices are incorrectly reporting connection status
+    // thus: deviceIsConnected?	{ deviceIsConnected: true }
+    // then if disconnecting [BleError: Device D7:D6:67:E0:02:34 is not connected]
+    // in which case an error is thrown when trying to connect: [BleError: Device ? is already connected]
+    // to work around this, we disconnect the device, ignoring any errors, before connecting again
```
Comment (josh-griffin) [inline .yalc/msupply-ble-service/src/Bluetooth/BleService.ts:86]: I mean...

😭

File: .yalc/msupply-ble-service/src/Bluetooth/BleService.ts [hunk 4]
```diff
@@ -79,9 +79,20 @@ export class BleService {
...
+    if (device.deviceType === BLUE_MAESTRO) {
+      try {
+        await this.manager.cancelDeviceConnection(device.id);
+      } catch {
+        // ignore error
+      }
+    } else {
+      const deviceIsConnected = await this.manager.isDeviceConnected(device.id);
+      this.logger.info('deviceIsConnected?', { deviceIsConnected });
       await this.manager.cancelDeviceConnection(device.id);
```
Comment (josh-griffin) [inline .yalc/msupply-ble-service/src/Bluetooth/BleService.ts:96]: we call `connectAndDiscoverServices` at the start of each of the commands.. I wonder if also cancelling the connection after each command, also, would somehow make things more stable 😅

**Review by josh-griffin — APPROVED** [review]
Haven't tested as I don't have a device but as crazy as it sounds, the changes make sense...? 😅

File: .yalc/msupply-ble-service/src/Bluetooth/BleService.ts (hunk 4 above)
Comment (mark-prins) [reply .yalc/msupply-ble-service/src/Bluetooth/BleService.ts:96]: good point - I half wondered about that. will have a try!

**Comment by mark-prins** [comment]
awesome - thanks @josh-griffin for taking a peek!

**Pull request merged**